### Classify the email using the binary classification method. Email Spam detection has two states: 
a) Normal State Not Spam

b) Abnormal State Spam. Use K-Nearest Neighbors 

and Support Vector Machine for classification. Analyze their performance. 
Dataset link: The emails.csv dataset on the Kaggle 
https://www.kaggle.com/datasets/balaka18/email-spam-classification-dataset-csv

In [4]:
import pandas as pd

In [5]:
df = pd.read_csv('https://raw.githubusercontent.com/Tanmay49/SEM7/main/ML/emails.csv')
df.head()

Email No.  the  to  ect  and  for  of    a  you  hou  ...  connevey  jay  \
0   Email 1    0   0    1    0    0   0    2    0    0  ...         0    0   
1   Email 2    8  13   24    6    6   2  102    1   27  ...         0    0   
2   Email 3    0   0    1    0    0   0    8    0    0  ...         0    0   
3   Email 4    0   5   22    0    5   1   51    2   10  ...         0    0   
4   Email 5    7   6   17    1    5   2   57    0    9  ...         0    0   

   valued  lay  infrastructure  military  allowing  ff  dry  Prediction  
0       0    0               0         0         0   0    0           0  
1       0    0               0         0         0   1    0           0  
2       0    0               0         0         0   0    0           0  
3       0    0               0         0         0   0    0           0  
4       0    0               0         0         0   1    0           0  

[5 rows x 3002 columns]

In [8]:
print(df['Prediction'].unique())

[0 1]


In [9]:
df.isna().sum()

Email No.     0
the           0
to            0
ect           0
and           0
             ..
military      0
allowing      0
ff            0
dry           0
Prediction    0
Length: 3002, dtype: int64

In [10]:
df.describe()

the           to          ect          and          for  \
count  5172.000000  5172.000000  5172.000000  5172.000000  5172.000000   
mean      6.640565     6.188128     5.143852     3.075599     3.124710   
std      11.745009     9.534576    14.101142     6.045970     4.680522   
min       0.000000     0.000000     1.000000     0.000000     0.000000   
25%       0.000000     1.000000     1.000000     0.000000     1.000000   
50%       3.000000     3.000000     1.000000     1.000000     2.000000   
75%       8.000000     7.000000     4.000000     3.000000     4.000000   
max     210.000000   132.000000   344.000000    89.000000    47.000000   

                of            a          you          hou           in  ...  \
count  5172.000000  5172.000000  5172.000000  5172.000000  5172.000000  ...   
mean      2.627030    55.517401     2.466551     2.024362    10.600155  ...   
std       6.229845    87.574172     4.314444     6.967878    19.281892  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000    12.000000     0.000000     0.000000     1.000000  ...   
50%       1.000000    28.000000     1.000000     0.000000     5.000000  ...   
75%       2.000000    62.250000     3.000000     1.000000    12.000000  ...   
max      77.000000  1898.000000    70.000000   167.000000   223.000000  ...   

          connevey          jay       valued          lay  infrastructure  \
count  5172.000000  5172.000000  5172.000000  5172.000000     5172.000000   
mean      0.005027     0.012568     0.010634     0.098028        0.004254   
std       0.105788     0.199682     0.116693     0.569532        0.096252   
min       0.000000     0.000000     0.000000     0.000000        0.000000   
25%       0.000000     0.000000     0.000000     0.000000        0.000000   
50%       0.000000     0.000000     0.000000     0.000000        0.000000   
75%       0.000000     0.000000     0.000000     0.000000        0.000000   
max       4.000000     7.000000     2.000000    12.000000        3.000000   

          military     allowing           ff          dry   Prediction  
count  5172.000000  5172.000000  5172.000000  5172.000000  5172.000000  
mean      0.006574     0.004060     0.914733     0.006961     0.290023  
std       0.138908     0.072145     2.780203     0.098086     0.453817  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     1.000000     0.000000     1.000000  
max       4.000000     3.000000   114.000000     4.000000     1.000000  

[8 rows x 3001 columns]

In [14]:
X = df.iloc[:, 1:3001]
Y = df.iloc[:, -1]  ## prediction column
X

the  to  ect  and  for  of    a  you  hou  in  ...  enhancements  \
0       0   0    1    0    0   0    2    0    0   0  ...             0   
1       8  13   24    6    6   2  102    1   27  18  ...             0   
2       0   0    1    0    0   0    8    0    0   4  ...             0   
3       0   5   22    0    5   1   51    2   10   1  ...             0   
4       7   6   17    1    5   2   57    0    9   3  ...             0   
...   ...  ..  ...  ...  ...  ..  ...  ...  ...  ..  ...           ...   
5167    2   2    2    3    0   0   32    0    0   5  ...             0   
5168   35  27   11    2    6   5  151    4    3  23  ...             0   
5169    0   0    1    1    0   0   11    0    0   1  ...             0   
5170    2   7    1    0    2   1   28    2    0   8  ...             0   
5171   22  24    5    1    6   5  148    8    2  23  ...             0   

      connevey  jay  valued  lay  infrastructure  military  allowing  ff  dry  
0            0    0       0    0               0         0         0   0    0  
1            0    0       0    0               0         0         0   1    0  
2            0    0       0    0               0         0         0   0    0  
3            0    0       0    0               0         0         0   0    0  
4            0    0       0    0               0         0         0   1    0  
...        ...  ...     ...  ...             ...       ...       ...  ..  ...  
5167         0    0       0    0               0         0         0   0    0  
5168         0    0       0    0               0         0         0   1    0  
5169         0    0       0    0               0         0         0   0    0  
5170         0    0       0    0               0         0         0   1    0  
5171         0    0       0    0               0         0         0   0    0  

[5172 rows x 3000 columns]

In [15]:
Y

0       0
1       0
2       0
3       0
4       0
       ..
5167    0
5168    0
5169    1
5170    1
5171    0
Name: Prediction, Length: 5172, dtype: int64

In [16]:
from sklearn.model_selection import train_test_split
# Split the data into training and testing sets
# x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.25, random_state=42)

x_train,x_test,y_train,y_test=train_test_split(X,Y)

In [17]:
from sklearn.neighbors import KNeighborsClassifier
model=KNeighborsClassifier()
model.fit(x_train,y_train)
y_pred=model.predict(x_test)

In [18]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.93      0.90      0.91       892
           1       0.79      0.86      0.82       401

    accuracy                           0.88      1293
   macro avg       0.86      0.88      0.87      1293
weighted avg       0.89      0.88      0.89      1293



In [19]:
from sklearn.svm import SVC ## note syntax
model=SVC()
model.fit(x_train,y_train)
y_pred=model.predict(x_test)

In [13]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.78      0.97      0.87       909
           1       0.84      0.37      0.51       384

    accuracy                           0.79      1293
   macro avg       0.81      0.67      0.69      1293
weighted avg       0.80      0.79      0.76      1293



# Analysis of the Classification Reports:
K-Nearest Neighbors (KNN):

High precision and recall for Class 0.
Class 1 has decent recall and slightly lower precision.
Overall, the accuracy is around 90%, showing that the model performs well for this dataset.
Support Vector Classifier (SVC):

Good precision for both classes, but recall for Class 1 is significantly lower (0.43).
This implies that the SVC struggles to correctly identify instances of Class 1, leading to a lower F1-score and poorer balance overall.
The accuracy is lower (82%), and the macro average F1-score is 0.73, indicating poorer performance, especially for Class 1.
Conclusion:
Ensure the training (model.fit) is only done on the training set (x_train, y_train) to maintain the integrity of the model evaluation.
The KNN model seems to perform better than the SVC for this dataset, as indicated by the higher F1-scores and better recall for Class 1.

"""
Explanation of Terms
Precision:
 The proportion of true positive predictions out of all positive predictions made.
Formula: 
Precision = (Tp)/(Tp+Fp)
 Indicates how many of the predicted positive cases were actually correct.

Recall (Sensitivity):

The proportion of true positive cases out of all actual positives.
Formula: 
Recall= (Tp)/(Tp+Fn)
Indicates how many actual positive cases were correctly predicted by the model.

F1-Score:

The harmonic mean of precision and recall.
Formula: 
2*precision×Recall
Provides a balance between precision and recall, useful when you need to find an equilibrium between the two.
Support:

The number of actual occurrences of each class in the dataset.
Indicates how many samples there were for each class.
Accuracy:

The proportion of correctly predicted samples out of the total samples.
Formula: 
Accuracy= 
Total Number of Samples/Number of Correct Predictions

 
Macro Average:

The average of precision, recall, and F1-score calculated independently for each class and then averaged.
Treats all classes equally, without considering the class imbalance.
Weighted Average:

Similar to the macro average, but it considers the number of instances (support) for each class when averaging.
Gives more weight to the scores of the classes with more samples.
Conclusion from the Output
Class 0 has a higher precision (0.95) and recall (0.91) compared to Class 1, indicating that the model performs better in predicting Class 0.
Class 1 has lower precision (0.80) but a relatively high recall (0.89), suggesting that while the model can find most of the actual positive samples for Class 1, it sometimes misclassifies negative samples as positive.
The overall accuracy of the model is 90%, which indicates that 90% of the total predictions were correct.
The macro average shows that both classes are fairly balanced in terms of overall performance with an average F1-score of 0.89.
The weighted average is also 0.91, showing that the average score is influenced by the number of samples in each class.
General Conclusion:
The model has good performance overall, particularly for Class 0, which is more common (917 samples). Class 1, although less frequent (376 samples), also shows reasonably good recall and F1-score, but with slightly lower precision.
"""